### PR TITLE
Use async open from OS

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

Fixes #

##  TODO

* [ ] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

File: Realm/Realm/Handles/ThreadSafeReferenceHandle.cs [hunk 1]
```diff
@@ -27,16 +27,30 @@ private static class NativeMethods
...
+
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "realm_thread_safe_reference_destroy", CallingConvention = CallingConvention.Cdecl)]
+            public static extern void destroy_realm_reference(IntPtr handle);
         }
 
+        private bool _isRealmReference;
+
         [Preserve]
-        public ThreadSafeReferenceHandle(IntPtr handle) : base(null, handle)
+        public ThreadSafeReferenceHandle(IntPtr handle, bool isRealmReference = false) : base(null, handle)
```
Comment (nirinchev) [inline Realm/Realm/Handles/ThreadSafeReferenceHandle.cs]: This is quite ugly, but couldn't find a better way to do it.

Comment (fealebenpae) [reply Realm/Realm/Handles/ThreadSafeReferenceHandle.cs]: It's going to go away with Core 6 soon, so we'll just live with it for a while.

File: Realm/Realm/Realm.cs [hunk 2]
```diff
@@ -93,7 +93,8 @@ public static Realm GetInstance(RealmConfigurationBase config = null)
         /// </remarks>
         /// <returns>A <see cref="Task{Realm}"/> that is completed once the remote realm is fully synchronized or immediately if it's a local realm.</returns>
         /// <param name="config">A configuration object that describes the realm.</param>
-        public static Task<Realm> GetInstanceAsync(RealmConfigurationBase config = null)
+        /// <param name="cancellationToken">An optional cancellation token that can be used to cancel the work.</param>
+        public static Task<Realm> GetInstanceAsync(RealmConfigurationBase config = null, CancellationToken? cancellationToken = null)
```
Comment (fealebenpae) [inline Realm/Realm/Realm.cs]:
```suggestion
        public static Task<Realm> GetInstanceAsync(RealmConfigurationBase config = null, CancellationToken cancellationToken = default)
```

File: Realm/Realm/Realm.cs [hunk 3]
```diff
@@ -114,7 +115,7 @@ public static Task<Realm> GetInstanceAsync(RealmConfigurationBase config = null)
                 schema = RealmSchema.Default;
             }
 
-            return config.CreateRealmAsync(schema);
+            return config.CreateRealmAsync(schema, cancellationToken ?? CancellationToken.None);
```
Comment (fealebenpae) [inline Realm/Realm/Realm.cs]:
```suggestion
            return config.CreateRealmAsync(schema, cancellationToken);
```
with the above suggestion

File: Realm/Realm/Realm.cs [hunk 4]
```diff
@@ -1170,4 +1171,4 @@ internal void DrainTransactionQueue()
             }
         }
     }
-}
```
Comment (fealebenpae) [inline Realm/Realm/Realm.cs]: newline

File: Realm/Realm/Server/Notifier/NotifierRealmConfiguration.cs [hunk 5]
```diff
@@ -41,9 +42,9 @@ internal override Realm CreateRealm(RealmSchema schema)
             return new Realm(handle, this, schema);
         }
 
-        internal override Task<Realm> CreateRealmAsync(RealmSchema schema)
+        internal override Task<Realm> CreateRealmAsync(RealmSchema schema, CancellationToken cancellationToken)
         {
             return Task.FromResult(CreateRealm(schema));
         }
     }
-}
```
Comment (fealebenpae) [inline Realm/Realm/Server/Notifier/NotifierRealmConfiguration.cs]: newline

File: Realm/Realm/Sync/ProgressNotifications/SyncProgress.cs [hunk 6]
```diff
@@ -43,5 +43,7 @@ internal SyncProgress(ulong transferred, ulong transferable)
             TransferredBytes = transferred;
             TransferableBytes = transferable;
         }
+
+        internal bool IsComplete => TransferableBytes == TransferredBytes;
     }
-}
```
Comment (fealebenpae) [inline Realm/Realm/Sync/ProgressNotifications/SyncProgress.cs]: newline

File: Realm/Realm/Sync/Session.cs [hunk 7]
```diff
@@ -217,4 +209,4 @@ internal void CloseHandle()
             }
         }
     }
-}
```
Comment (fealebenpae) [inline Realm/Realm/Sync/Session.cs]: newline

File: Tests/Realm.Tests/Server/ShouldHandleTests.cs [hunk 8]
```diff
@@ -68,11 +67,11 @@ public void ShouldHandle_WhenRealmExists_InvokedOnStart()
...
                                                                          retryDelay: 100,
-                                                                         attempts: 10); // 1 second
+                                                                         attempts: 100); // 1 second
 
                     Assert.True(containsNewRealm);
                 }
             }, timeout: 30000);
         }
     }
-}
```
Comment (fealebenpae) [inline Tests/Realm.Tests/Server/ShouldHandleTests.cs]: newline

File: Tests/Realm.Tests/Sync/SyncTestBase.cs [hunk 9]
```diff
@@ -106,12 +107,12 @@ protected Realm GetRealm(RealmConfigurationBase config)
             return result;
         }
 
-        protected async Task<Realm> GetRealmAsync(RealmConfigurationBase config, bool openAsync = true)
+        protected async Task<Realm> GetRealmAsync(RealmConfigurationBase config, bool openAsync = true, CancellationToken? cancellationToken = null)
```
Comment (fealebenpae) [inline Tests/Realm.Tests/Sync/SyncTestBase.cs]:
```suggestion
        protected async Task<Realm> GetRealmAsync(RealmConfigurationBase config, bool openAsync = true, CancellationToken cancellationToken = default)
```

File: Tests/Realm.Tests/Sync/SyncTestBase.cs [hunk 10]
```diff
@@ -123,4 +124,4 @@ protected async Task<Realm> GetRealmAsync(RealmConfigurationBase config, bool op
             return result;
         }
     }
-}
```
Comment (fealebenpae) [inline Tests/Realm.Tests/Sync/SyncTestBase.cs]: newline

File: Tests/Realm.Tests/Sync/SynchronizedInstanceTests.cs [hunk 11]
```diff
@@ -178,45 +179,60 @@ public void GetInstanceAsync_ReportsProgress()
...
+        public void GetInstanceAsync_Cancel_ShouldCancelWait()
+        {
+            SyncTestHelpers.RunRosTestAsync(async () =>
+            {
+                var config = await SyncTestHelpers.GetIntegrationConfigAsync("foo");
+                await PopulateData(config);
+                // Update config to make sure we're not opening the same Realm file.
+                config = new FullSyncConfiguration(config.ServerUri, config.User, config.DatabasePath + "1");
+
+                var cts = new CancellationTokenSource();
```
Comment (fealebenpae) [inline Tests/Realm.Tests/Sync/SynchronizedInstanceTests.cs]: This should be disposed.

File: wrappers/src/async_open_task_cs.cpp [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+REALM_EXPORT void realm_async_open_task_unregister_progress_notifier(const SharedAsyncOpenTask& task, uint64_t token, NativeException::Marshallable& ex)
+{
+	return handle_errors(ex, [&] {
+		task->unregister_download_progress_notifier(token);
+	});
+}
+
+}
```
Comment (fealebenpae) [inline wrappers/src/async_open_task_cs.cpp]: newline

Comment (fealebenpae) [reply wrappers/src/async_open_task_cs.cpp]: Also this file appears to be using tabs instead of spaces for indentation.

**Pull request merged**